Add support for column numbers (and a few fixes)

boris-kolpackov opened the pull request:
We have tested this against a fuzzer as well as with line/column tests in our own code (https://github.com/libstud/libstud-json/blob/master/tests/parser/basics/basics.testscript).

francoisk committed:
Fix few issues with position tracking

- Position was not being tracked in case of a user-provided get function.
- It was possible for position to get incremented beyond end of input.

francoisk committed:
Add support for column numbers

francoisk committed:
Remove superfluous semicolon

boris-kolpackov commented:
Ping.

skissane commented:
I think this is a good PR overall, but I've noticed a couple of behaviours with the column numbers which seem slightly odd:

* Attempting to read an empty JSON document reports an error at line 1 column 0 instead of at line 1 column 1.
* Given a string or numeric literal, the column number points to its start but the byte position points to its end

Both of the above behaviours seem counterintuitive to me, and I personally think this PR would be better if they were both changed (report error at column 1 not 0 at start of empty document; have column number and byte position of literal point to the same place, either both to the start or both to the end). But I don't know how widely my intuitions are shared.

boris-kolpackov commented:
Thanks for reveiwing/testing this.

> Attempting to read an empty JSON document reports an error at line 1 column 0 instead of at line 1 column 1.

Yes, good catch. I think this tweak to `json_get_column()` should do the trick:

```
size_t json_get_column(json_stream *json)
{
    return json->colno == 0
               ? json->source.position == 0 ? 1 : json->source.position - json->linepos - json->lineadj
               : json->colno;
}
```

> Given a string or numeric literal, the column number points to its start but the byte position points to its end

While this is not stated explicitly in the documentation (though alluded to by saying that it's always available), the position tracking is not specific to error handling but is rather a generic current byte position in the input. In fact, from the code it is clear this is the position of the next character to be read as opposed to the last character that contributed to the current event. I think changing it to always point to the first character that contributed to the current event could close the door for other (than diagnostics) use-case (and could break existing code that relied on the current semantics).

I think if something like this is desirable, we should rather invent a separate function with this semantics, for example, `json_get_error_position()` or, better yet, `json_get_event_position()`. In this case it probably makes sense to adjust the current semantics of `json_get_column()` to match `json_get_position()` and to provide `json_get_event_column()`. I don't believe we would need  `json_get_event_line()` since an even cannot span multiple input lines but we could provide one for symmetry.

skissane commented:
@boris-kolpackov I tested your proposed change to `json_get_column`, it indeed fixes the issue about column 0.

Regarding the other issue, I think it would make sense if there were two different sets of functions, one to return current parsing position (line/col/pos) and the other to return the start of the last event (line/col/pos). I guess if there has been no event yet, it just returns line=1,col=1,pos=0.

About "I don't believe we would need json_get_event_line()" => suppose I have a JSON literal followed by some newlines. It is possible the parsing position is some way into those newlines, especially if `json_source_get` has been called. ~~Or, maybe I have a JSON literal, followed by some newlines, then a syntax error – the last event line/col/pos might be still at the start of that literal, while the current parsing line/col/pos is at the error location.~~ **(Edit: actually that can't happen because the syntax error would be an Error event so the last event location and current parsing location would be the same)** So I think the last event line number and the current parsing position line number could be different, which suggests two different functions would make sense.

boris-kolpackov committed:
Handle empty document special case in json_get_column()

Specifically, we should return 1:1 (line:column) instead of 1:0. See
GitHub skeeto/pdjson/pull/26 for details.

boris-kolpackov commented:
> Yes, good catch. I think this tweak to json_get_column() should do the trick:

FWIW, I've applied the fix to [my fork of pdjson](https://github.com/boris-kolpackov/pdjson). I am happy to update this PR if/when there is intent to merge it.

boris-kolpackov committed:
Fix bug in read_utf8() function

Specifically, the comparison to EOF should be performed before casting
the value to char.

boris-kolpackov committed:
Make test/stream.c print diagnostics to stderr in case of error

boris-kolpackov committed:
Don't attempt to print EOF characters in diagnostics

boris-kolpackov commented:
FYI, the first two commits in this PR have now been applied to master.
